pkg/reconciler/apiserversource/apiserversource.go

```diff
@@ -433,3 +450,70 @@ func (r *Reconciler) UpdateFromMetricsConfigMap(cfg *corev1.ConfigMap) {
 	}
 	logging.FromContext(r.loggingContext).Info("Update from metrics ConfigMap", zap.Any("ConfigMap", cfg))
 }
+
+func (r *Reconciler) runAccessCheck(src *v1alpha1.ApiServerSource) error {
+	if src.Spec.Resources == nil || len(src.Spec.Resources) == 0 {
+		src.Status.MarkSufficientPermissions()
+		return nil
+	}
+
+	user := "system:serviceaccount:" + src.Namespace + ":"
+	if src.Spec.ServiceAccountName == "" {
+		user += "default"
+	} else {
+		user += src.Spec.ServiceAccountName
+	}
+
+	verbs := []string{"get", "list", "watch"}
+	resources := src.Spec.Resources
+	lastReason := ""
+
+	// Collect all missing permissions.
+	missing := ""
+	sep := ""
+
+	for _, res := range resources {
+		gv, err := schema.ParseGroupVersion(res.APIVersion)
+		if err != nil { // shouldn't happened after #2134 is fixed
+			return err
+		}
+		gvr, _ := meta.UnsafeGuessKindToResource(schema.GroupVersionKind{Kind: res.Kind, Group: gv.Group, Version: gv.Version})
+		missingVerbs := ""
+		sep1 := ""
+		for _, verb := range verbs {
+			sar := &authorizationv1.SubjectAccessReview{
+				Spec: authorizationv1.SubjectAccessReviewSpec{
+					ResourceAttributes: &authorizationv1.ResourceAttributes{
+						Namespace: src.Namespace,
+						Verb:      verb,
+						Group:     gv.Group,
+						Resource:  gvr.Resource,
+					},
+					User: user,
+				},
+			}
+
+			response, err := r.KubeClientSet.AuthorizationV1().SubjectAccessReviews().Create(sar)
+			if err != nil {
+				return err
+			}
+
+			if !response.Status.Allowed {
+				missingVerbs += sep1 + verb
+				sep1 = ", "
+			}
+		}
+		if missingVerbs != "" {
```

Comment:
Just a thought. If we had a more structured error, it might make things like testing easier? Say you would return PermissionsError that would wrap into something like
map[string][]string
so that you could then return tuples of missing permissions? Just a thought.

Reply:
I would rather keep it the way it is just because PersmissionsError does not help that much for testing and it's only helping for testing. If you feel strongly about it, let me know and I'll make the change.